src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+  return promise_result::await(web_usb_device.call<val>("open")).error;
+}
+```
+
+### Device enumeration
+
+Of course, before any device can be opened, libusb needs to retrieve a list of available devices. The backend must implement this operation via a `get_device_list` handler.
+
+The difficulty is that, unlike on other platforms, there is no way to enumerate all the connected USB devices on the web for security reasons. Instead, the flow is split into two parts. First, the web application can request devices with specific properties via [`navigator.usb.requestDevice()`](https://developer.mozilla.org/docs/Web/API/USB/requestDevice) and the user will manually choose which device they want to expose or they can reject the permission prompt. Then, you can list the already approved and connected devices via [`navigator.usb.getDevices()`](https://developer.mozilla.org/docs/Web/API/USB/getDevices).
+
+At first I tried to use `requestDevice()` in the implementation of `get_device_list` handler to make usage easier. However, showing a permission prompt with a list of connected devices is considered a sensitive operation, and it must be triggered by user interaction (like a button click on a page), otherwise it always returns a rejected promise. libusb applications might often want to list the connected devices upon application start-up, so using `requestDevice()` was not an option.
+
+Instead, I had to leave invocation of `navigator.usb.requestDevice()` to the end developer, and only expose the already approved devices from `navigator.usb.getDevices()`:
+
+```cpp
+// Store the global `navigator.usb` once upon initialisation.
+thread_local const val web_usb = val::global("navigator")["usb"];
+
+int em_get_device_list(libusb_context *ctx, discovered_devs **devs) {
+  // C++ equivalent of `await navigator.usb.getDevices()`.
+  // Note: at this point we must already have some devices exposed -
+  // caller must have called `await navigator.usb.requestDevice(...)`
+  // in response to user interaction before going to LibUSB.
+  // Otherwise this list will be empty.
+  auto result = promise_result::await(web_usb.call<val>("getDevices"));
+  if (result.error) {
+    return result.error;
+  }
+  auto &web_usb_devices = result.value;
+  // Iterate over the exposed devices.
+  uint8_t devices_num = web_usb_devices["length"].as<uint8_t>();
+  for (uint8_t i = 0; i < devices_num; i++) {
+    auto web_usb_device = web_usb_devices[i];
+    // …
+    *devs = discovered_devs_append(*devs, dev);
+  }
+  return LIBUSB_SUCCESS;
+}
+```
+
+Most of the backend code uses `val` and `promise_result` in a similar way as already shown above. There were few more interesting hacks I've had to do to convert properties exposed by WebUSB to those expected by libusb, both in this method and in others handling data transfer, but those implementation details are less important for the purposes of this article. Make sure to check the code and the comments if you're interested.
```

Comment:
```suggestion
Most of the backend code uses `val` and `promise_result` in a similar way to what's shown above. There were few more interesting hacks I've had to do to convert properties exposed by WebUSB to those expected by libusb, both in this method and in others handling data transfer, but those implementation details are less important for the purposes of this article. Make sure to check the code and the comments if you're interested.
```

Reply:
I can't work what the middle sentence is saying. Please try something else. This problem usually comes from trying to say too much in a single sentence. Try breaking it up.

Reply:
```suggestion
Most of the backend code uses `val` and `promise_result` in a similar way as already shown above. There are few more interesting hacks in the data transfer handling code, but those implementation details are less important for the purposes of this article. Make sure to check the code and comments on Github if you're interested.
```